packages/tunnel-client/src/lib/tunnel.ts

```diff
@@ -266,53 +228,63 @@ export class Tunnel extends (EventEmitter as new () => TypedEmitter<TunnelEvents
     http2Connections,
   }: Omit<TunnelInfo, "multiplexingRemotePort"> & {
     multiplexingRemotePort: number;
-  }): Promise<TunnelHTTP2Cluster> {
+  }): Promise<void> {
     const localProtocol = localHttps ? "https" : "http";
     this.logger.debug(
-      "establishing tunnel %s://%s:%s <> %s:%s",
+      "establishing tunnel %s://%s:%s <> %s:%s using cluster workers",
       localProtocol,
       localHost,
       localPort,
       remoteHost,
       multiplexingRemotePort,
     );
 
-    const commonTunnelOpts = {
-      logger: this.logger,
-      localHost,
-      localPort,
-      localHttps,
-      allowInvalidCert,
-      proxyAllUrls,
-      rewriteHostnameToAppUrl,
-      enableDnsCache,
-    };
+    const numWorkers = http2Connections || DEFAULT_HTTP2_NUMBER_OF_CONNECTIONS;
+    const workerPath = path.resolve(__dirname, "./worker-entrypoint.js");
 
-    const numConnections =
-      http2Connections ?? DEFAULT_HTTP2_NUMBER_OF_CONNECTIONS;
-    const sockets = await Promise.all(
-      Array.from({ length: numConnections }).map(() =>
-        this.openSocket({
-          useTls,
-          remoteHost,
-          multiplexingRemotePort,
-          tunnelPassphrase,
-          sendAuthOkAck: true,
-        }),
-      ),
-    );
+    cluster.setupPrimary({
+      exec: workerPath,
+      silent: false,
+      execArgv: ["--max-old-space-size=2048"],
+    });
+
+    for (let i = 1; i <= numWorkers; i++) {
+      const workerOptions: WorkerInitOptions = {
+        workerId: i,
+        useTls,
+        remoteHost,
+        multiplexingRemotePort,
+        tunnelPassphrase,
+        localHost,
+        localPort,
+        localHttps,
+        ...(localCert ? { localCert } : {}),
+        ...(localKey ? { localKey } : {}),
+        ...(localCa ? { localCa } : {}),
+        allowInvalidCert,
+        proxyAllUrls,
+        rewriteHostnameToAppUrl,
+        enableDnsCache,
+      };
+
+      const worker = cluster.fork({
+        WORKER_OPTIONS: JSON.stringify(workerOptions),
```

Comment:
I think we can probably pass these as messages to improve the security of this given that we're the socket passphrase.